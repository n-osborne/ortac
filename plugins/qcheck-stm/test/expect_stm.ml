open Fmt
open Ortac_qcheck_stm

let _ =
  let open Reserr in
  let* sigs, config = Config.init "lib.mli" "make 'a' 42" "(char, int) t" in
  let* ir = Ir_of_gospel.run sigs config in
  let cmd = Stm_of_ir.cmd_type config ir in
  let state = Stm_of_ir.state_type ir in
  pf stdout "%a@." Ppxlib_ast.Pprintast.structure [ cmd; state ];
  ok ()
