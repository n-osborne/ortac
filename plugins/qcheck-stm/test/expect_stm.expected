type nonrec cmd =
  | Length 
  | Pop 
  | Push of char 
  | Extend 
  | To_list [@@deriving show { with_path = false }]
type nonrec state = {
  size: int ;
  contents: char list }
let arb_cmd _ =
  let open QCheck in
    make ~print:show_cmd
      (let open Gen in
         [pure Length;
         pure Pop;
         (pure (fun a_1 -> Push a_1)) <*> char;
         pure Extend;
         pure To_list])
let next_state cmd__001_ state__002_ =
  match cmd__001_ with
  | Length -> state__002_
  | Pop ->
      {
        state__002_ with
        contents = (Ortac_runtime.Gospelstdlib.List.tl state__002_.contents)
      }
  | Push a_1 ->
      {
        state__002_ with
        contents =
          (if
             (Ortac_runtime.Gospelstdlib.List.length state__002_.contents) =
               state__002_.size
           then state__002_.contents
           else a_1 :: state__002_.contents)
      }
  | Extend ->
      {
        state__002_ with
        size =
          (Ortac_runtime.Gospelstdlib.( * )
             (Ortac_runtime.Gospelstdlib.integer_of_int 2) state__002_.size);
        contents = []
      }
  | To_list -> state__002_
let precond cmd__007_ state__008_ =
  match cmd__007_ with
  | Length -> true
  | Pop ->
      Ortac_runtime.Gospelstdlib.(>)
        (Ortac_runtime.Gospelstdlib.List.length state__008_.contents)
        (Ortac_runtime.Gospelstdlib.integer_of_int 0)
  | Push a_1 -> true
  | Extend -> true
  | To_list -> true
let postcond cmd__003_ state__004_ res__005_ =
  let new_state__006_ = lazy (next_state cmd__003_ state__004_) in
  match (cmd__003_, res__005_) with
  | (Length, Res ((Int, _), i)) ->
      (Ortac_runtime.Gospelstdlib.integer_of_int i) =
        (Ortac_runtime.Gospelstdlib.List.length
           (Lazy.force new_state__006_).contents)
  | (Pop, Res ((Char, _), a_2)) ->
      a_2 = (Ortac_runtime.Gospelstdlib.List.hd state__004_.contents)
  | (Push a_1, Res ((Unit, _), _)) -> true
  | (Extend, Res ((Unit, _), _)) -> true
  | (To_list, Res ((List (Char), _), l)) ->
      l = (Lazy.force new_state__006_).contents
let run cmd__009_ sut__010_ =
  match cmd__009_ with
  | Length -> Res (int, (length sut__010_))
  | Pop -> Res (char, (pop sut__010_))
  | Push a_1 -> Res (unit, (push a_1 sut__010_))
  | Extend -> Res (unit, (extend sut__010_))
  | To_list -> Res ((list char), (to_list sut__010_))
